File: samples/macOS/cpp/NonTableManagerSamples/CloseAllPositions/source/SampleParams.cpp
#include "stdafx.h"
#include <limits>
#include "SampleParams.h"

const char *SampleParams::Strings::instrumentNotSpecified = "'Instrument' is not specified (/i|-i|/instrument|--instrument)";
const char *SampleParams::Strings::buysellNotSpecified = "'BuySell' is not specified (/d|-d|/buysell|--buysell)";
const char *SampleParams::Strings::rateNotSpecified = "'Rate' is not specified (/r|-r|/rate|--rate)";
const char *SampleParams::Strings::rangeNotSpecified = "'Range' is not specified (/range|--range)";
const char *SampleParams::Strings::trailStepNotSpecified = "'TrailStep' is not specified (/trailstep|--trailstep)";
const char *SampleParams::Strings::ratestopNotSpecified = "'RateStop' is not specified (/ratestop|--ratestop)";
const char *SampleParams::Strings::ratelimitNotSpecified = "'RateLimit' is not specified (/ratelimit|--ratelimit)";
const char *SampleParams::Strings::orderidNotSpecified = "'OrderID' is not specified (/orderid|--orderid)";
const char *SampleParams::Strings::primaryidNotSpecified = "'PrimaryID' is not specified (/primaryid|--primaryid)";
const char *SampleParams::Strings::secondaryidNotSpecified = "'SecondaryID' is not specified (/secondaryid|--secondaryid)";
const char *SampleParams::Strings::contingencyidNotSpecified = "'ContingencyID' is not specified (/contingencyid|--contingencyid)";
const char *SampleParams::Strings::timeframeNotSpecified = "'Timeframe' is not specified (/timeframe|--timeframe)";
const char *SampleParams::Strings::statusNotSpecified = "'SubscriptionStatus' is not specified (/status|--status)";
const char *SampleParams::Strings::expDateNotSpecified = "'expDate' is not specified (/status|--status)";

SampleParams::SampleParams(int argc, char **argv)
{
    /* Load parameters with short keys. */
    mInstrument = getArgument(argc, argv, "i");
    mBuySell = getArgument(argc, argv, "d");

    std::string sRate = getArgument(argc, argv, "r");

    /* If parameters with short keys not loaded, load with long keys. */
    if (mInstrument.empty())
        mInstrument = getArgument(argc, argv, "instrument");
    if (mBuySell.empty())
        mBuySell = getArgument(argc, argv, "buysell");
    if (sRate.empty())
        sRate = getArgument(argc, argv, "rate");

    /* Load parameters with long keys. */
    mContingencyID = getArgument(argc, argv, "contingencyid");
    mOrderID = getArgument(argc, argv, "orderid");
    mPrimaryID = getArgument(argc, argv, "primaryid");
    mSecondaryID = getArgument(argc, argv, "secondaryid");
    mTimeframe = getArgument(argc, argv, "timeframe");
    mAccount = getArgument(argc, argv, "account");
    mOrderType = getArgument(argc, argv, "ordertype");
    mStatus = getArgument(argc, argv, "status");
    mExpDate = getArgument(argc, argv, "expireDate");

    std::string sLots = getArgument(argc, argv, "lots");
    std::string sDateFrom = getArgument(argc, argv, "datefrom");
    std::string sDateTo = getArgument(argc, argv, "dateto");
    std::string sRangeInPips = getArgument(argc, argv, "range");
    std::string sRateStop = getArgument(argc, argv, "ratestop");
    std::string sRateLimit = getArgument(argc, argv, "ratelimit");
    std::string sTrailStep = getArgument(argc, argv, "trailstep");

    /* Convert types. */
    if (sLots.empty())
        mLots = 1;
    else
        mLots = atoi(sLots.c_str());

    double const NaN = std::numeric_limits<double>::quiet_NaN();

    if (sRangeInPips.empty())
        mRangeInPips = NaN;
    else
        mRangeInPips = atoi(sRangeInPips.c_str());

    if (sTrailStep.empty())
        mTrailStep = NaN;
    else
        mTrailStep = atoi(sTrailStep.c_str());

    if (sRate.empty())
        mRate = NaN;
    else
        mRate = atof(sRate.c_str());

    if (sRangeInPips.empty())
        mRangeInPips = NaN;
    else
        mRangeInPips = atof(sRangeInPips.c_str());

    if (sRateStop.empty())
        mRateStop = NaN;
    else
        mRateStop = atof(sRateStop.c_str());

    if (sRateLimit.empty())
        mRateLimit = NaN;
    else
        mRateLimit = atof(sRateLimit.c_str());

    struct tm tmBuf = {0};

    if (sDateFrom.empty())
        mDateFrom = NaN;
    else
    {
        strptime(sDateFrom.c_str(), "%m.%d.%Y %H:%M:%S", &tmBuf);
        CO2GDateUtils::CTimeToOleTime(&tmBuf, &mDateFrom);
    }

    if (sDateTo.empty())
        mDateTo = NaN;
    else
    {
        strptime(sDateTo.c_str(), "%m.%d.%Y %H:%M:%S", &tmBuf);
        CO2GDateUtils::CTimeToOleTime(&tmBuf, &mDateTo);
    }

}

SampleParams::~SampleParams(void)
{
}

const char *SampleParams::getArgument(int argc, char **argv, const char *key)
{
    for (int i = 1; i < argc; ++i)
    {
        if (argv[i][0] == '-' || argv[i][0] == '/')
        {
            int iDelimOffset = 0;
            if (strncmp(argv[i], "--", 2) == 0)
                iDelimOffset = 2;
            else if (strncmp(argv[i], "-", 1) == 0 || strncmp(argv[i], "/", 1) == 0)
                iDelimOffset = 1;

            if (_stricmp(argv[i] + iDelimOffset, key) == 0 && argc > i+1)
                return argv[i+1];
        }
    }
    return "";
}

/** Getters. */

const char *SampleParams::getExpDate()
{
    return mExpDate.c_str();
}

const char *SampleParams::getInstrument()
{
    return mInstrument.c_str();
}

const char *SampleParams::getBuySell()
{
    return mBuySell.c_str();
}

const char *SampleParams::getContingencyID()
{
    return mContingencyID.c_str();
}

const char *SampleParams::getOrderID()
{
    return mOrderID.c_str();
}

const char *SampleParams::getPrimaryID()
{
    return mPrimaryID.c_str();
}

const char *SampleParams::getSecondaryID()
{
    return mSecondaryID.c_str();
}

const char *SampleParams::getTimeframe()
{
    return mTimeframe.c_str();
}

const char *SampleParams::getAccount()
{
    return mAccount.c_str();
}

const char *SampleParams::getOrderType()
{
    return mOrderType.c_str();
}

const char *SampleParams::getStatus()
{
    return mStatus.c_str();
}

int SampleParams::getLots()
{
    return mLots;
}

int SampleParams::getTrailStep()
{
    return mTrailStep;
}

DATE SampleParams::getDateFrom()
{
    return mDateFrom;
}

DATE SampleParams::getDateTo()
{
    return mDateTo;
}

double SampleParams::getRangeInPips()
{
    return mRangeInPips;
}

double SampleParams::getRate()
{
    return mRate;
}

double SampleParams::getRateStop()
{
    return mRateStop;
}

double SampleParams::getRateLimit()
{
    return mRateLimit;
}

/** Setters. */

void SampleParams::setAccount(const char *value)
{
    if (!value)
        mAccount = "";
    else
        mAccount = value;
}

void SampleParams::setOrderType(const char *value)
{
    if (!value)
        mOrderType = "";
    else
        mOrderType = value;
}

void SampleParams::setDateFrom(DATE value)
{
    mDateFrom = value;
}

void SampleParams::setDateTo(DATE value)
{
    mDateTo = value;
}

File: samples/Windows/java/TableManagerSamples/CalculateTradingCommissions/src/SampleParams.java
package common;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Calendar;

public class SampleParams {
    public static final String INSTRUMENT_NOT_SPECIFIED = "'Instrument' is not specified (/i|-i|/instrument|--instrument)";
    public static final String BUYSELL_NOT_SPECIFIED = "'BuySell' is not specified (/d|-d|/buysell|--buysell)";  

    // Getters

    public String getInstrument() {
        return mInstrument;
    }
    private String mInstrument;

    public String getBuySell() {
        return mBuySell;
    }
    private String mBuySell;

    public int getLots() {
        return mLots;
    }
    private int mLots;

    public String getAccountID() {
        return mAccountID;
    }
    private String mAccountID;

    public String getOrderType() {
        return mOrderType;
    }
    private String mOrderType;

    public Calendar getDateFrom() {
        return mDateFrom;
    }
    private Calendar mDateFrom;

    public Calendar getDateTo() {
        return mDateTo;
    }
    private Calendar mDateTo;

    public double getRate() {
        return mRate;
    }
    private double mRate;

    public double getRateStop() {
        return mRateStop;
    }
    private double mRateStop;

    public double getRateLimit() {
        return mRateLimit;
    }
    private double mRateLimit;

    public String getStatus() {
        return mStatus;
    }
    private String mStatus;

    public String getExpireDate() {
        return mExpireDate;
    }
    private String mExpireDate;
    // Setters

    public void setAccountID(String sAccountID) {
        mAccountID = sAccountID;
    }

    public void setOrderType(String sOrderType) {
        mOrderType = sOrderType;
    }

    public void setDateFrom(Calendar dtFrom) {
        mDateFrom = dtFrom;
    }

    public void setDateTo(Calendar dtTo) {
        mDateTo = dtTo;
    }

    // ctor
    public SampleParams(String[] args) {
        // Get parameters with short keys
        mInstrument = getArgument(args, "i");
        mBuySell = getArgument(args, "d");
        String sRate = getArgument(args, "r");

        // If parameters with short keys are not specified, get parameters with long keys
        if (mInstrument.isEmpty())
            mInstrument = getArgument(args, "instrument");
        if (mBuySell.isEmpty())
            mBuySell = getArgument(args, "buysell");
        if (sRate.isEmpty())
            sRate = getArgument(args, "rate");

        String sLots = "";
        String sDateFrom = "";
        String sDateTo = "";
        String sRateStop = "";
        String sRateLimit = "";

        // Get parameters with long keys
        sLots = getArgument(args, "lots");
        mAccountID = getArgument(args, "account");    

        // Convert types
        try {
            mLots = Integer.parseInt(sLots);
        } catch (Exception ex) {
            mLots = 1;
        }
        
    }

    private String getArgument(String[] args, String sKey) {
        for (int i = 0; i < args.length; i++) {
            int iDelimOffset = 0;
            if (args[i].startsWith("--")) {
                iDelimOffset = 2;
            } else if (args[i].startsWith("-") || args[i].startsWith("/")) {
                iDelimOffset = 1;
            }

            if (args[i].substring(iDelimOffset).equals(sKey) && (args.length > i+1)) {
                return args[i+1];
            }
        }
        return "";
    }
}
